llvm-external-projects/iree-dialects/lib/Dialect/LinalgExt/IR/LinalgExtOps.cpp

```diff
@@ -1447,6 +1447,118 @@ LogicalResult TopkOp::getResultTilePosition(
   return success();
 }
 
+//===----------------------------------------------------------------------===//
+// PackOp
+//===----------------------------------------------------------------------===//
+
+LogicalResult PackOp::verify() {
+  // input rank = r
+  // output rank = 2 * r (input rank + tile rank).
+  // inner tiles = r
+  return success();
+}
+
+SmallVector<OpFoldResult> PackOp::getMixedInnerTiles() {
+  SmallVector<OpFoldResult> mixedInnerTiles;
+  mixedInnerTiles.reserve(getInputRank());
+  unsigned dynamicValIndex = 0;
+  for (Attribute attr : getStaticInnerTiles()) {
+    auto tileAttr = attr.cast<IntegerAttr>();
+    if (!ShapedType::isDynamic(tileAttr.getInt()))
+      mixedInnerTiles.push_back(tileAttr);
+    else
+      mixedInnerTiles.push_back(getInnerTiles()[dynamicValIndex++]);
+  }
+  return mixedInnerTiles;
+}
+
+SmallVector<StringRef> PackOp::getLoopIteratorTypes() {
+  SmallVector<StringRef> iteratorTypes(getInputRank() * 2,
+                                       getParallelIteratorTypeName());
+  return iteratorTypes;
+}
+
+// copied from:
+// https://mlir.llvm.org/doxygen/TensorInferTypeOpInterfaceImpl_8cpp_source.html
+/// Helper function to convert a vector of `OpFoldResult`s into a vector of
+/// `Value`s.
+static SmallVector<Value> getAsValues(OpBuilder &b, Location loc,
```

Comment:
This should live in `StaticValueUtils.h` upstream

Reply:
I will open an upstream PR soon.